### PR TITLE
MULE-13226: Allow substitutionGroup as XMLHint in SDK's annotations

### PR DESCRIPTION
**Pull request by lucianoRM**

### CODE REVIEW

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/ObjectTypeSchemaDelegate.java [hunk 1]
```diff
@@ -410,9 +413,27 @@ private TopLevelElement registerAbstractElement(QName typeQName, DslElementSynta
...
 
+  private Optional<QName> resolveSubstitutionGroupFromString(String userConfiguredSubstitutionGroup) {
+    String[] splittedSubstitutionGroup = userConfiguredSubstitutionGroup.split(":");
+    if (splittedSubstitutionGroup.length == 2) {
+      String namespacePrefix = splittedSubstitutionGroup[0];
+      String substitutionComponent = splittedSubstitutionGroup[1];
+      String namespaceUri = builder.getNamespaceUri(namespacePrefix);
+      if (namespaceUri != null) {
+        return Optional.of(new QName(namespaceUri, substitutionComponent, namespacePrefix));
+      }
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/ObjectTypeSchemaDelegate.java]: else should it fail? @alegmarra

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/ObjectTypeSchemaDelegate.java [hunk 2]
```diff
@@ -410,9 +413,27 @@ private TopLevelElement registerAbstractElement(QName typeQName, DslElementSynta
...
+    return Optional.empty();
+  }
+
   private QName getAbstractElementSubstitutionGroup(DslElementSyntax typeDsl, Optional<DslElementSyntax> baseDsl) {
     QName substitutionGroup;
-    if (baseDsl.isPresent()) {
+    //First check if the substitutionGroup was defined by the user
+    if (!typeDsl.getSubstitutionGroup().isEmpty()) {
+      substitutionGroup = resolveSubstitutionGroupFromString(typeDsl.getSubstitutionGroup())
+          .orElseThrow(() -> new IllegalArgumentException(typeDsl.getSubstitutionGroup()
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/ObjectTypeSchemaDelegate.java]: resolveSubstitutionGroupFromString() should fail instead of leaving it up to the consumer to do so

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/ObjectTypeSchemaDelegate.java (hunk 1 above)
Comment (lucianoRM) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/ObjectTypeSchemaDelegate.java]: Means that the prefix is neither "mule" nor "ee" and that it's not within the imported namespaces.

Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/ObjectTypeSchemaDelegate.java]: i know what i means. What I'm saying is that it should fail right here. Look at the following comment

**Comment by lucianoRM** [comment]
Last changes do not include requests in comment. Working on it...

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/ObjectTypeSchemaDelegate.java (hunk 1 above)
Comment (lucianoRM) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/ObjectTypeSchemaDelegate.java]: I thought that the question was if it was ok to fail under those conditions.

File: modules/extensions-spring-support/pom.xml [hunk 3]
```diff
@@ -208,6 +208,13 @@
         </dependency>
         <dependency>
             <groupId>org.mule.tests</groupId>
+            <artifactId>mule-substitution-group-extension</artifactId>
+            <version>4.0.0-SNAPSHOT</version>
```
Comment (alegmarra) [inline modules/extensions-spring-support/pom.xml]: ${project.version}

File: tests/test-extensions/substitution-group-extension/src/main/java/org/mule/test/substitutiongroup/extension/InsignificantPojo.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.test.substitutiongroup.extension;
+
+public class InsignificantPojo {
```
Comment (alegmarra) [inline tests/test-extensions/substitution-group-extension/src/main/java/org/mule/test/substitutiongroup/extension/InsignificantPojo.java]: why this?

File: tests/test-extensions/substitution-group-extension/pom.xml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    <packaging>jar</packaging>
+
+    <name>Mule Substitution Group Test Extension</name>
+    <description>A Mule extension for testing</description>
+    <dependencies>
+        <dependency>
+            <groupId>org.mule.tests</groupId>
+            <artifactId>mule-heisenberg-extension</artifactId>
+            <version>4.0.0-SNAPSHOT</version>
```
Comment (alegmarra) [inline tests/test-extensions/substitution-group-extension/pom.xml]: ${project.version}

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/ObjectTypeSchemaDelegate.java [hunk 6]
```diff
@@ -410,6 +419,16 @@ private TopLevelElement registerAbstractElement(QName typeQName, DslElementSynta
     return abstractElement;
   }
 
+  private QName resolveSubstitutionGroup(SubstitutionGroup userConfiguredSubstitutionGroup) {
+    String namespaceUri = builder.getNamespaceUri(userConfiguredSubstitutionGroup.getPrefix());
```
Comment (alegmarra) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/ObjectTypeSchemaDelegate.java]: is better to put all this logic in a `builder.resolveSubstitutionGroup` and not having the QName being crafted in the delegate.

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/SchemaBuilder.java [hunk 7]
```diff
@@ -706,4 +707,17 @@ void addInlineParameterGroup(ParameterGroupModel group, ExplicitGroup parentSequ
...
+
+  String getNamespaceUri(String prefix) {
+    if (prefix.equals(CORE_PREFIX))
+      return CORE_NAMESPACE;
+    if (prefix.equals(EE_PREFIX))
+      return EE_NAMESPACE;
+    Optional<ExtensionModel> extensionModelFromPrefix = dslContext.getExtensions().stream()
+        .filter((extensionModel) -> prefix.equals(extensionModel.getXmlDslModel().getPrefix())).findFirst();
+    if (extensionModelFromPrefix.isPresent()) {
+      return extensionModelFromPrefix.get().getXmlDslModel().getNamespace();
```
Comment (alegmarra) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/capability/xml/schema/builder/SchemaBuilder.java]: this is not right, we never check that the namespace returned was actually imported by the schema. The "ee" namespace is not being imported in the schema either (check the xsd). 
- import the namespace if it was not imported before 
- add a test that validates this imports 
- verify at least manually that the substitutiongroup.xsd is a valid schema. Better yet to actually use the extension in some xml that gets parsed and executed by mule

File: modules/extensions-spring-support/src/test/resources/models/heisenberg.json [hunk 8]
```diff
@@ -16115,10 +16115,9 @@
                         "parent": "",
                         "genericTypes": []
                       },
-                      "xmlHints": {
+                      "typeXmlHints": {
```
Comment (alegmarra) [inline modules/extensions-spring-support/src/test/resources/models/heisenberg.json]: is this rename actually necessary? is there a clash with any other type annotation?

Comment (lucianoRM) [inline modules/extensions-spring-support/src/test/resources/models/heisenberg.json]: it's just to be consistent with the interface name
